bench/bm_buffer.cpp

```diff
@@ -0,0 +1,153 @@
+                BufferWriter auto writer = buffer.new_writer();
+                std::size_t nSamplesProduced = 0;
+                barrier.arrive_and_wait();
+                while (nSamplesProduced <= (min_samples + nProducer - 1) / nProducer) {
+                    writer.publish([](auto &) {}, vector_length);
+                    nSamplesProduced += vector_length;
+                }
+                barrier.arrive_and_wait();
+            }
+        });
+    }
+
+    // init consumers
+    std::vector<std::thread> consumers;
+    for (int i = 0; i < nConsumer; i++) {
+        consumers.emplace_back([&]() {
+            setCpuAffinity(threadID++);
+            barrier.arrive_and_wait();
+            for (int rep = 0; rep < n_repeat; ++rep) {
+                BufferReader auto reader = buffer.new_reader();
+                std::size_t nSamplesConsumed = 0;
+                barrier.arrive_and_wait();
+                while (nSamplesConsumed < min_samples) {
+                    if (reader.available() < vector_length) {
+                        continue;
+                    }
+                    const auto &input = reader.get(vector_length);
+                    nSamplesConsumed += input.size();
+
+                    if (!reader.consume(input.size())) {
+                        throw std::runtime_error(fmt::format("could not consume {} samples", input.size()));
+                    }
+                }
+                barrier.arrive_and_wait();
+            }
+        });
+    }
+
+    // all producers and consumer are ready, waiting to give the sign
+    barrier.arrive_and_wait();
+    ::benchmark::benchmark<n_repeat>(fmt::format("{}: {} writers -<{}>-> {} readers", name, nProducer, vector_length,
+                                                 nConsumer),
+                                     min_samples)
+            = [&]() {
+                  barrier.arrive_and_wait();
+                  barrier.arrive_and_wait();
+              };
+
+    // clean up
+    for (std::thread &thread : producers) thread.join();
+    for (std::thread &thread : consumers) thread.join();
+}
+
+inline const boost::ut::suite _buffer_tests = [] {
+    const uint64_t samples = 100'000'000; // minimum number of samples
+
+    for (int nP = 1; nP <= 4; nP *= 2) {
+        for (int nR = 1; nR <= 4; nR *= 2) {
+            if (not (nP == 1 and nR == 1)) {
+                benchmark::results::add_separator();
+            }
+            for (int buffer_type = 4; buffer_type <= 5; ++buffer_type) {
+                for (int veclen = 16; veclen <= 4096; veclen *= 16) { // vector length (<-> size of bulk operation)
+                    const std::size_t size = std::max(4096, veclen) * nR * 10;
+                    auto allocator = std::pmr::polymorphic_allocator<int32_t>();
+                    auto invoke = [&](auto buffer) {
+                        testNewAPI(buffer, veclen, samples, nP, nR, buffer_type == 4 ? "POSIX" : "portable");
```

Comment:
I like that both cases are benchmarked against each other the POSIX/mmap double-mapped buffer has been used in GR for ages/historic reason but the benchmark pretty much shows that the results are either equivalent or the portable/memcpy-based version even faster in some cases. :+1:

Reply:
Hmm, for `veclen == 16` the "POSIX" vs. "portable" results are the same order of magnitude, but for `veclen == 4096` there is a factor 10-20 throughput difference (your screenshot above, but that's also what I saw here).

Reply:
I would not interpret too much into my screenshot which was done in a hurry. Anything beyond 10 GS/s is anyway more L1/L2/L3 cache related and can become very quickly flaky on what the OS/processor is doing in parallel, temperature, pressure, cosmic background radiation (esp. for the 100 GS/s results), ...